Rework SpeziSecureStorage API

lukaskollmer opened the pull request:
# Rework SpeziSecureStorage API

## :recycle: Current situation & Problem
- SpeziStorage currently consists of two targets: SpeziLocalStorage and SpeziSecureStorage
  - The names are misleading; the fact that one of them is called the "secure" storage might lead people to assume that the other one is not secure. This is false. They both are.
  - → we rename `SecureStorage` to `CredentialsStorage`, to make its name match more closely its intended purpose, and also get rid of any potential for misunderstanding the individual modules' intended purposes
- The fact that the `CredentialsStorage` API is using raw `String` values to identify entries in the keychain, and is making heavy use of optional parameters with default `nil` values, makes it very easy to mess up operations that are intended to operate on the same entry in the keychain (since you need to remember to always pass in the exact same set of parameters in all your callsites, which is very easy to get wrong)
  - in fact, there are some bugs in the current (ie, pre-PR) implementation which are a result of nil-defaulted function parameters not being specified when they should be given an explicit value.
  - → we rework the API a little bit, to introduce dedicated `KeyTag` (for working with keys) and `CredentialsStorageKey` (for working with credentials) types, which are used to identify, in a more safe manner, the entries an operation should work with.
  - The `CredentialsStorageKey`, for example, combines all information which is required to correctly define how a credentials entry is / should be stored in the keychain, instead of requiring the user to remember to always pass in the correct same values whenever they want to store/retrieve/delete a credentials entry.


## :gear: Release Notes 
- Renamed and reworked `SecureStorage` into `CredentialsStorage`
- Introduced `KeyTag` type to identify keys
- Introduced `CredentialsStorageKey` type to define credentials entries in the `CredentialsStorage`.


## :books: Documentation
The documentation has been updated to reflect the new structure of the API.


## :white_check_mark: Testing
The tests have been updated to work with the new API.
Since the functionality of the API is unchanged, there are no new tests.


## :pencil: Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md).

lukaskollmer committed:
rename `SpeziSecureStorage` → `SpeziCredentialsStorage`

why? both the SecureStorage and the LocalStorage are in fact secure (unless you intentially instruct the LocalStorage not to encrypt smth), but calling one the "secure" storage kinda makes it sound like the other one is not secure. which is wrong. they both are.

lukaskollmer committed:
import sorting

lukaskollmer committed:
introduce a dedicated `KeyTag` type

lukaskollmer committed:
x

lukaskollmer committed:
rework SecureStorage → CredentialsStorage

lukaskollmer committed:
improve docs, cleanup

lukaskollmer committed:
update docs

lukaskollmer committed:
readme

PSchmiedmayer commented on .spi.yml:
```diff
@@ -12,4 +12,4 @@ builder:
     - platform: ios
       documentation_targets:
       - SpeziLocalStorage
```
Wondering if we should rename this to `SpeziFileStorage`?

lukaskollmer replied:
i'd be against that, tbh. i also thought about potential new names for the LocalStorage target, but i wasn't able to come up with anything good.

calling it "file storage" makes it sound like the module just does file storage management, whereas what it's actually doing is data storage management, and the fact that the data are stored as files is merely an implementation detail (which is the way it should be, IMO)

PSchmiedmayer replied:
Sounds fair; thank you for the input!

PSchmiedmayer commented on .spi.yml:
```diff
@@ -12,4 +12,4 @@ builder:
       documentation_targets:
       - SpeziLocalStorage
-      - SpeziSecureStorage
+      - SpeziCredentialsStorage
```
I like the new name; both of them are secure by default anyways. File/Local storage is using encryption by default anyways.

PSchmiedmayer commented on README.md:
```diff
@@ -66,47 +66,34 @@ struct ExampleStorageView: View {
 ### Handling Keys
 
-Similar to `Credentials` instances, you can also use the `SecureStorage` module to interact with keys.
+Similar to `Credentials` instances, you can also use the `CredentialsStorage` module to interact with keys.
```
Wondering if this should even be moved to a separate target, just to be consistent and clear with the naming? And maybe add `cryptographic` to ensure the people understand what we are dealing with here?
```suggestion
Similar to `Credentials` instances, you can also use the `CredentialsStorage` module to interact with cryptographic keys.
```

lukaskollmer replied:
since (cryptographic) keys can also be thought of as a kind of credential, i'd say it's fine to keep it in the same module.

agree w/ the wording suggestion 👍

PSchmiedmayer commented on Sources/SpeziCredentialsStorage/CredentialsStorageKey.swift:
````diff
@@ -0,0 +1,59 @@
+///     // ...
+/// }
+/// ```
+public struct CredentialsStorageKey: Hashable, Sendable {
````
Nice mechanism!

lukaskollmer replied:
btw, i intentionally decided not to write an equivalent of the `@LocalStorageEntry` here, since a) it's (AFAIK) impossible to observe keychain changes, and b) the `CredentialsStorageKey` isn't really a 100% equivalent to the `LocalStorageKey`, so we also can't use the mechanism of locking all accesses over the key itself.

PSchmiedmayer replied:
Agree, makes sense to me.

PSchmiedmayer commented on Sources/SpeziCredentialsStorage/CredentialsStorageScope.swift:
```diff
@@ -31,14 +33,6 @@ public enum SecureStorageScope: Equatable, Identifiable {
     case keychainSynchronizable(accessGroup: String? = nil)
     
     
-    /// Store the element in the Secure Enclave
-    public static let secureEnclave = secureEnclave()
-    /// Store the element in the Keychain
-    public static let keychain = keychain()
-    /// Store the element in the Keychain and enable it to be synchronizable between different instances of user devices.
-    public static let keychainSynchronizable = keychainSynchronizable()
```
Wondering what the reason behind removing the enum like elements was? Sometimes it is kind of nice and easy to explore if we just have enum cases without the need to write parentheses here.

lukaskollmer replied:
there were some places in the code that were doing an `assert(!(scope ~= .secureEnclave))` check, which compiled but didn't do what it was supposed to (it just checked against the default (no user presence) secureEnclave case (ie, the static property), rather than checking against the `secureEnclave` case in general.

i absolutely agree that it looks nicer, and we could bring these back if you want to, but i'm kinda wary in general that e.g. if someone does `if scope == .keychain` they think they're checking for the overall keychain case, rather than comparing against the more specific `keychain` static property

PSchmiedmayer replied:
Hmm, could we provide a Equatable implementation (or `~=`) that covers these cases to rather implement this on this level?

lukaskollmer replied:
i'll see if i can change it back in a way that makes it work as one would expect while also maintaining the nicer syntax 👍

PSchmiedmayer commented on Sources/SpeziCredentialsStorage/SpeziCredentialsStorage.docc/SpeziCredentialsStorage.md:
Thank you for improving and updating all the docs!

PSchmiedmayer reviewed (approved):
Thank you for working on this! I had some smaller comments, that would be great to be addressed. The major question is if we should separate credentials and keys in separate targets with some shared foundations?

lukaskollmer committed:
Update README.md

Co-authored-by: Paul Schmiedmayer <PSchmiedmayer@users.noreply.github.com>

philippzagar reviewed (approved):
Thanks @lukaskollmer for the great second PR in the Storage module! 🚀 
Liked the nice test cases as well as the docs!

PSchmiedmayer commented on README.md:
```diff
@@ -66,47 +66,34 @@ struct ExampleStorageView: View {
 
 
 ### Handling Keys
 
-Similar to `Credentials` instances, you can also use the `SecureStorage` module to interact with keys.
+Similar to `Credentials` instances, you can also use the `CredentialsStorage` module to interact with cryptographic keys.
```
I am wondering if we should introduce an abstraction layer for the keychain interactions (just internal to the package, e.g. exposed with `package`) and then have a `CredentialsStorage` and `KeyStorage` that are separate targets & types?

lukaskollmer replied:
i really like this idea. it'd also allow us to easily add the ability to allow apps to store certificates using `CredentialsStorage` (since these are just another form of keychain item).

we might even want to make the keychain abstraction layer public[1] in case apps need that level of raw access / need to go beyond the level of control the CredentialsStorage provides.

[1] similar to like how with SpeziBluetooth and SpeziDevices we both have a package that implements the low-level BT interactions, plus one that give you a couple of well-known device interactions and UI, if that's sufficient for your needs

PSchmiedmayer replied:
Agree; that would be great! And yes, it might even make sense to make that public as well, good point.

lukaskollmer committed:
bring back the storage scope static properties for

lukaskollmer committed:
adjust line length

lukaskollmer committed:
KeychainStorage

lukaskollmer committed:
doca

lukaskollmer committed:
swiftlint

lukaskollmer committed:
prefer using CFString as a dictionary key

lukaskollmer commented:
@pauljohanneskraft just to loop you in on this (@PSchmiedmayer told me that you're working on the android side of things on this): i'm currently in the process of reworking the SpeziStorage API a little bit. We first updated the LocalStorage module in #30 (to move away from using string-based keys and having implicit parameters, to instead using explicit storage keys to better define and identify items in the storage, and prevent a couple of potential accidental mistakes when using the API). This PR does the same for the SpeziSecureStorage API (which is now called SpeziKeychainStorage).

lukaskollmer commented:
@PSchmiedmayer i reworked the now-former SecureStorage module a bit, based on your feedback: it's now called KeychainStorage, since i figured this would be a good overall naming scheme (having it called LocalStorage and KeychainStorage means that both modules, via their name, directly tell you where your data will end up, also also indirectly whoch kinds of data each module is intended for)

I rewrote pretty much the entire implementation here (the public API is mostly identical in terms of which operations we support, except that some things have been renamed, and there's additional APIs for retrieving/deleting items).

There were a couple things which previously were broken/incorrect[^1] which should now work as expected
- cryptographic keys were always created in the secure enclave, regardless of which storage scope (now called storage option) or access group was specified
- retrieving or deleting synchronizable keys or credentials would always fail (returning nil), because the `kSecAttrSynchronizable` option wasn't set in the query
- `SecAccesControl` objects were created and used only for synchronizable items, which seems to be one kind of item that doesn't support them
- for items that are both synchronizable and have some kind of access restriction (which isn't necessarily the same as `SecAccesControl`), the access restriction wasn't applied

I also ended up adding/exposing some additional functionality: we now support both generic and internet passwords, have dedicated types for these, and also make it easier to work with cryptographic keys, via eg computed properties that provide access to individual attributes on Credentials or Keys, new bulk-retriveal functions (which exist mainly to make it easier for apps to manage their keychain items).


[^1] i *think* that they were incorrect; i've spent the last couple of days doing a deep dive into the keychain API and i'm still like only 85% sure i understand how all of the different options and flags integrate with each other

lukaskollmer committed:
switch XCTestExtensions dependency to GH

PSchmiedmayer commented on Sources/SpeziKeychainStorage/Credentials/Credentials.swift:
Very cool; these extensions are impressive and comprehensive!

PSchmiedmayer reviewed (approved):
Amazing job @lukaskollmer; thank you for all the context in the PR!

Looks really good, let me know if I should merge it with admin permissions.

lukaskollmer commented:
@PSchmiedmayer in terms of features, the PR is complete, but i do wanna get the test coverage up a bit before we merge it (there seems to be some issue currently bc it's saying that none of the new code is tested, which is wrong (there are tests))

PSchmiedmayer commented:
Sounds great; thank you!

lukaskollmer committed:
XCTestExtensions dependency

lukaskollmer committed:
undo accidentally commiting my bundle id

lukaskollmer committed:
try to fix the code coverage >.<

lukaskollmer committed:
fix tag-based credentials query not taking `isSynchronizable` into account

lukaskollmer committed:
Credentials: fix CFBoolean-typed attributes not being processed correctly

for some reason comparing against kCFBooleanTrue (which is what the docs say the value should be) doesn't work. (seems like the value is actually a CFNumber???)

lukaskollmer committed:
Credentials: fix hashing implementation

lukaskollmer committed:
update tests

lukaskollmer committed:
AdJUsT LinE LEnGhT

lukaskollmer committed:
swiftlint

lukaskollmer committed:
switch XCTestExtensions dependency back to a tagged-release version

lukaskollmer committed:
some more rework

lukaskollmer committed:
ugh

lukaskollmer committed:
move the `deleteAll...` functions behind an `Internal` SPI

lukaskollmer committed:
hmmm

lukaskollmer committed:
pure desperation

lukaskollmer committed:
revert commit

lukaskollmer committed:
include KeyTag in key creation error

lukaskollmer committed:
hmmm

lukaskollmer committed:
lint

lukaskollmer committed:
test more things

lukaskollmer committed:
uff

lukaskollmer committed:
test more stuff

lukaskollmer committed:
fix SecKey extensions not working for synchronized keys

(the issue was that `extendedAttributes` would always be an empty dict for such keys, since the SecItemCopyMatching operation would always fail because it wasn't able to find the key

lukaskollmer committed:
expand tests

lukaskollmer committed:
happy linter happy life

lukaskollmer committed:
moar test coverage

lukaskollmer committed:
x

lukaskollmer committed:
x

lukaskollmer committed:
even moar
